src/types/api.js

```diff
@@ -17,6 +17,32 @@
  * @prop {string} text
  * @prop {string} uri
  * @prop {string} user
+ * @prop {boolean} hidden
+ * @prop {boolean} [$highlight]
+ * @prop {boolean} [$orphan]
+ * @prop {boolean} [$anchorTimeout]
+ *
+ * @prop {Object} document
+ *   @prop {string} document.title
+ *
+ * @prop {Object} permissions
+ *   @prop {string[]} permissions.read
+ *   @prop {string[]} permissions.update
+ *   @prop {string[]} permissions.delete
+ *
+ * @prop {Object[]} target
+ *   @prop {string} target.source
+ *   @prop {Object[]} target.selector
+ *     @prop {string} target.selector.type
+ *     @prop {string} target.selector.exact
+ *     @prop {number} target.selector.start
```

Comment:
The properties that a selector has depends on the type of the selector. Creating a type with all the properties and marking them optional is one way of representing this. A better way though is to use a _union_:

```js
/**
 * @typedef TextQuoteSelector
 * @prop {'TextQuoteSelector'} type
 * @prop {string} exact
 * @prop {string} [prefix]
 * @prop {string} [suffix]
 */

/**
 * @typedef TextPositionSelector
 * @prop {'TextPositionSelector'} type
 * @prop {number} start
 * @prop {number} end
 */

/**
 * @typedef {TextPositionSelector|TextQuoteSelector} Selector
 */
```

The advantage of this is that it provides better documentation to the reader, and it also enables TS to do
more accurate checking that you're accessing props which exist:

```js
const selector = ...
if (selector.type === 'TextQuoteSelector') {
  // TS knows that `selector` is a `TextQuoteSelector` here and
  // has the properties "exact", "suffix" and "prefix".
}
```

Reply:
That makes a lot more sense, I was just bliding filling props as I saw them used here and there. One advantage of doing this TS documentation is we now have a concrete.definition for the shit we're using!